pkg/apis/serving/k8s_validation.go

```diff
@@ -267,7 +276,51 @@ func ValidatePodSpec(ps corev1.PodSpec) *apis.FieldError {
 	return errs
 }
 
+//ValidateMultiContainerPorts validates port when specified multiple containers
+func ValidateMultiContainerPorts(containers []corev1.Container) *apis.FieldError {
+	cPort := []int32{}
+	var errs *apis.FieldError
+	for i := range containers {
+		for j := range containers[i].Ports {
+			cPort = append(cPort, containers[i].Ports[j].ContainerPort)
+		}
+	}
+	if len(cPort) == 0 {
+		errs = errs.Also(apis.ErrMissingField("containers.ports.containerPort"))
+	}
+	if len(cPort) > 1 {
+		errs = errs.Also(apis.ErrMultipleOneOf("containers.ports.containerPort"))
+	}
+	return errs
+}
+
+//ValidateMultiContainer validate fields for non serving containers
+func ValidateMultiContainer(containers *corev1.Container, volumes sets.String) *apis.FieldError {
+	var errs *apis.FieldError
+	validateMultiContainerProbe := func(p *corev1.Probe) *apis.FieldError {
+		if p != nil {
+			return apis.CheckDisallowedFields(*p, *ProbeMask(&corev1.Probe{}))
+		}
+		return nil
+	}
+	errs = errs.Also(validateMultiContainerProbe(containers.LivenessProbe).ViaField("livenessProbe"))
+	errs = errs.Also(validateMultiContainerProbe(containers.ReadinessProbe).ViaField("readinessProbe"))
+	errs = errs.Also(validate(*containers, volumes))
+	return errs
+}
+
+//ValidateContainer validate fields for serving containers
 func ValidateContainer(container corev1.Container, volumes sets.String) *apis.FieldError {
+	var errs *apis.FieldError
+	// Liveness Probes
+	errs = errs.Also(validateProbe(container.LivenessProbe).ViaField("livenessProbe"))
+	// Readiness Probes
+	errs = errs.Also(validateReadinessProbe(container.ReadinessProbe).ViaField("readinessProbe"))
+	errs = errs.Also(validate(container, volumes))
+	return errs
+}
+
+func validate(container corev1.Container, volumes sets.String) *apis.FieldError {
 	if equality.Semantic.DeepEqual(container, corev1.Container{}) {
 		return apis.ErrMissingField(apis.CurrentField)
 	}
```

Comment:
Related to this change, but tangential: we should expand the list of reservedContainerNames in case we need to add more system sidecars later. Perhaps something with a prefix of `knative-serving-system-*`?

This is potentially breaking, but given that this requires named containers means that the risk of breaking only goes up after this lands.

Reply:
It doesn't require named containers, does it? Aren't we inferring names from the indices? @savitaashture

Reply:
Yes if user don't provide name for sidecar container we inferring names from the indices
https://github.com/knative/serving/pull/6739/files#diff-daca68b8123c709e8127c6c834c5c023R56